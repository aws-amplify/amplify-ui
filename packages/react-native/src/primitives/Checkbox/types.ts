import {
  ImageStyle,
  PressableProps,
  StyleProp,
  TextStyle,
  ViewStyle,
} from 'react-native';
import { IconProps } from '../Icon/types';
import { LabelProps } from '../Label';
import { LabelPosition } from '../Label/types';

export interface CheckboxProps<T> extends PressableProps {
  iconStyle?: IconProps['style'];
  label?: string;
  labelPosition?: LabelPosition;
  labelStyle?: LabelProps['style'];
  onChange: (value: T) => void;
  selected?: boolean;
  size?: IconProps['size'];
  style?: StyleProp<ViewStyle>;
  value: T;
}

export interface CheckboxStyles {
  container?: ViewStyle;
  disabled?: ViewStyle;
  icon?: ImageStyle;
  label?: TextStyle;
}
