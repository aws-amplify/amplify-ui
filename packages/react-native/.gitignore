dist
.watchmanconfig

# ignore jest generated coverage directory
coverage 
