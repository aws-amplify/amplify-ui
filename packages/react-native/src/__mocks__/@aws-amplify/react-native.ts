export default {
  loadUrlPolyfill: jest.fn(),
};
