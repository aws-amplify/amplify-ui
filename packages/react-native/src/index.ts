export {
  Authenticator,
  AuthenticatorProps,
  useAuthenticator,
  UseAuthenticator,
  withAuthenticator,
  WithAuthenticatorOptions,
} from './Authenticator';
export * from './InAppMessaging';
export {
  Theme,
  ThemeProvider,
  defaultDarkModeOverride,
  defaultTheme,
  useTheme,
} from './theme';
