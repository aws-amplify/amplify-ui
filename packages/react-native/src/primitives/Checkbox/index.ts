export { default as Checkbox } from './Checkbox';
export type { CheckboxProps, CheckboxStyles } from './types';
