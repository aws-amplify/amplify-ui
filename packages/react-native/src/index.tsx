import React from 'react';
import { Text, View } from 'react-native';

export function HelloWorld(): JSX.Element {
  return (
    <View>
      <Text>Hello World!</Text>
    </View>
  );
}
