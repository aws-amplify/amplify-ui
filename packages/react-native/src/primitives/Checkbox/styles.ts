import { StyleSheet } from 'react-native';

import { StrictTheme } from '../../theme';
import { CheckboxStyles } from './types';
import { getFlexDirectionFromLabelPosition } from '../Label/utils';
import { LabelPosition } from '../Label/types';

export const getThemedStyles = (
  theme: StrictTheme,
  labelPosition: LabelPosition
): CheckboxStyles => {
  const { tokens, components } = theme;
  const { opacities, space } = tokens;

  return StyleSheet.create({
    container: {
      alignItems: 'center',
      padding: space.xs,
      flexDirection: getFlexDirectionFromLabelPosition(labelPosition),
      ...components?.checkbox?.container,
    },
    disabled: {
      opacity: opacities[60],
      ...components?.checkbox?.disabled,
    },
    icon: {
      ...components?.checkbox?.icon,
    },
    label: {
      ...components?.checkbox?.label,
    },
  });
};
