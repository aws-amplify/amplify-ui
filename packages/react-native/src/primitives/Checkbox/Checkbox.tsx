import React, { useCallback, useMemo, useState } from 'react';
import { Pressable, ViewStyle } from 'react-native';

import { CheckboxProps } from './types';
import { Icon } from '../Icon';
import { Label } from '../Label';
import { getThemedStyles } from './styles';
import { icons } from '../../assets';
import { useTheme } from '../../theme';

export default function Checkbox<T>({
  accessibilityRole = 'checkbox',
  iconStyle,
  disabled,
  label,
  labelPosition = 'end',
  labelStyle,
  onChange,
  selected,
  size,
  style,
  value,
  ...rest
}: CheckboxProps<T>): JSX.Element {
  const [checked, setChecked] = useState(selected ?? false);
  const theme = useTheme();
  const themedStyle = getThemedStyles(theme, labelPosition);

  const handleOnChange = useCallback(() => {
    onChange?.(value);
    setChecked(!checked);
  }, [onChange, value, checked]);

  const containerStyle: ViewStyle = useMemo(
    () => ({
      ...themedStyle.container,
      ...(disabled && themedStyle.disabled),
    }),
    [disabled, themedStyle]
  );

  return (
    <Pressable
      {...rest}
      accessibilityRole={accessibilityRole}
      disabled={disabled}
      onPress={handleOnChange}
      style={[containerStyle, style]}
    >
      <Icon
        source={checked ? icons.checkboxFilled : icons.checkboxOutline}
        size={size}
        style={iconStyle}
      />
      {label ? <Label style={labelStyle}>{label}</Label> : null}
    </Pressable>
  );
}
