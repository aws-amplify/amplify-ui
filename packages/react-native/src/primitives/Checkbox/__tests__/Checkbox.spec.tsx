import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';
import Checkbox from '../Checkbox';

const onChange = jest.fn();

describe('Checkbox', () => {
  beforeEach(() => {
    onChange.mockClear();
  });

  [true, false].forEach((value) => {
    it(`renders as expected when selected is ${value}`, () => {
      const { toJSON } = render(
        <Checkbox selected={value} value={value} onChange={onChange} />
      );
      expect(toJSON()).toMatchSnapshot();
    });
  });

  it('renders as expected when disabled', () => {
    const { toJSON } = render(
      <Checkbox disabled value="" onChange={onChange} />
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders as expected with accessibilityRole', () => {
    const { toJSON } = render(
      <Checkbox value="" onChange={onChange} accessibilityRole={'none'} />
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('calls the expected handler when selected', () => {
    const { getByRole } = render(<Checkbox value="" onChange={onChange} />);
    const checkbox = getByRole('checkbox');
    fireEvent.press(checkbox);
    expect(onChange).toHaveBeenCalledTimes(1);
  });

  it('does nothing when disabled', () => {
    const { getByRole } = render(
      <Checkbox value="" onChange={onChange} disabled />
    );
    const checkbox = getByRole('checkbox');
    fireEvent.press(checkbox);
    expect(onChange).not.toHaveBeenCalled();
  });
});
