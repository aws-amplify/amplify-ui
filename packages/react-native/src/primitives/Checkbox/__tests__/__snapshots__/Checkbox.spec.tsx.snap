// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox applies label position prop 1`] = `
<View
  accessibilityRole="checkbox"
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "column",
        "padding": 8,
      },
      undefined,
    ]
  }
  testID="checkboxTestID"
>
  <Image
    accessibilityRole="image"
    source={
      Object {
        "testUri": "../../../src/assets/icons/checkboxOutline.png",
      }
    }
    style={
      Array [
        Object {
          "height": 24,
          "resizeMode": "contain",
          "tintColor": undefined,
          "width": 24,
        },
        undefined,
      ]
    }
  />
  <Text
    accessibilityRole="text"
    style={
      Array [
        Object {
          "fontSize": 16,
          "fontWeight": "400",
        },
        Object {
          "color": "hsl(210, 50%, 10%)",
        },
        undefined,
      ]
    }
  >
    test
  </Text>
</View>
`;

exports[`Checkbox applies theme and style props 1`] = `
<View
  accessibilityRole="checkbox"
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "padding": 8,
      },
      Object {
        "padding": 10,
      },
    ]
  }
  testID="checkboxTestID"
>
  <Image
    accessibilityRole="image"
    source={
      Object {
        "testUri": "../../../src/assets/icons/checkboxOutline.png",
      }
    }
    style={
      Array [
        Object {
          "height": 24,
          "resizeMode": "contain",
          "tintColor": undefined,
          "width": 24,
        },
        Object {},
      ]
    }
  />
  <Text
    accessibilityRole="text"
    style={
      Array [
        Object {
          "fontSize": 16,
          "fontWeight": "400",
        },
        Object {
          "color": "hsl(210, 50%, 10%)",
        },
        Object {},
      ]
    }
  >
    test
  </Text>
</View>
`;

exports[`Checkbox renders as expected when disabled 1`] = `
<View
  accessibilityRole="checkbox"
  accessibilityState={
    Object {
      "disabled": true,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "opacity": 0.6,
        "padding": 8,
      },
      undefined,
    ]
  }
  testID="checkboxTestID"
>
  <Image
    accessibilityRole="image"
    source={
      Object {
        "testUri": "../../../src/assets/icons/checkboxOutline.png",
      }
    }
    style={
      Array [
        Object {
          "height": 24,
          "resizeMode": "contain",
          "tintColor": undefined,
          "width": 24,
        },
        undefined,
      ]
    }
  />
  <Text
    accessibilityRole="text"
    style={
      Array [
        Object {
          "fontSize": 16,
          "fontWeight": "400",
        },
        Object {
          "color": "hsl(210, 50%, 10%)",
        },
        undefined,
      ]
    }
  >
    test
  </Text>
</View>
`;

exports[`Checkbox renders as expected when passing a number to the size prop 1`] = `
<View
  accessibilityRole="checkbox"
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "padding": 8,
      },
      undefined,
    ]
  }
  testID="checkboxTestID"
>
  <Image
    accessibilityRole="image"
    source={
      Object {
        "testUri": "../../../src/assets/icons/checkboxOutline.png",
      }
    }
    style={
      Array [
        Object {
          "height": 42,
          "resizeMode": "contain",
          "tintColor": undefined,
          "width": 42,
        },
        undefined,
      ]
    }
  />
  <Text
    accessibilityRole="text"
    style={
      Array [
        Object {
          "fontSize": 16,
          "fontWeight": "400",
        },
        Object {
          "color": "hsl(210, 50%, 10%)",
        },
        undefined,
      ]
    }
  >
    test
  </Text>
</View>
`;

exports[`Checkbox renders as expected when selected is false 1`] = `
<View
  accessibilityRole="checkbox"
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "padding": 8,
      },
      undefined,
    ]
  }
  testID="checkboxTestID"
>
  <Image
    accessibilityRole="image"
    source={
      Object {
        "testUri": "../../../src/assets/icons/checkboxOutline.png",
      }
    }
    style={
      Array [
        Object {
          "height": 24,
          "resizeMode": "contain",
          "tintColor": undefined,
          "width": 24,
        },
        undefined,
      ]
    }
  />
  <Text
    accessibilityRole="text"
    style={
      Array [
        Object {
          "fontSize": 16,
          "fontWeight": "400",
        },
        Object {
          "color": "hsl(210, 50%, 10%)",
        },
        undefined,
      ]
    }
  >
    test
  </Text>
</View>
`;

exports[`Checkbox renders as expected when selected is true 1`] = `
<View
  accessibilityRole="checkbox"
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "padding": 8,
      },
      undefined,
    ]
  }
  testID="checkboxTestID"
>
  <Image
    accessibilityRole="image"
    source={
      Object {
        "testUri": "../../../src/assets/icons/checkboxFilled.png",
      }
    }
    style={
      Array [
        Object {
          "height": 24,
          "resizeMode": "contain",
          "tintColor": undefined,
          "width": 24,
        },
        undefined,
      ]
    }
  />
  <Text
    accessibilityRole="text"
    style={
      Array [
        Object {
          "fontSize": 16,
          "fontWeight": "400",
        },
        Object {
          "color": "hsl(210, 50%, 10%)",
        },
        undefined,
      ]
    }
  >
    test
  </Text>
</View>
`;

exports[`Checkbox renders as expected with custom accessibilityRole 1`] = `
<View
  accessibilityRole="none"
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "padding": 8,
      },
      undefined,
    ]
  }
  testID="checkboxTestID"
>
  <Image
    accessibilityRole="image"
    source={
      Object {
        "testUri": "../../../src/assets/icons/checkboxOutline.png",
      }
    }
    style={
      Array [
        Object {
          "height": 24,
          "resizeMode": "contain",
          "tintColor": undefined,
          "width": 24,
        },
        undefined,
      ]
    }
  />
  <Text
    accessibilityRole="text"
    style={
      Array [
        Object {
          "fontSize": 16,
          "fontWeight": "400",
        },
        Object {
          "color": "hsl(210, 50%, 10%)",
        },
        undefined,
      ]
    }
  >
    test
  </Text>
</View>
`;

exports[`Checkbox renders as expected without a label 1`] = `
<View
  accessibilityRole="checkbox"
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "padding": 8,
      },
      undefined,
    ]
  }
  testID="checkboxTestID"
>
  <Image
    accessibilityRole="image"
    source={
      Object {
        "testUri": "../../../src/assets/icons/checkboxOutline.png",
      }
    }
    style={
      Array [
        Object {
          "height": 24,
          "resizeMode": "contain",
          "tintColor": undefined,
          "width": 24,
        },
        undefined,
      ]
    }
  />
</View>
`;
